New version

EVALORE opened the pull request.

EVALORE committed:
s1

EVALORE committed:
s2

EVALORE committed:
s3

EVALORE committed:
s4

EVALORE committed:
s5

EVALORE committed:
s6

EVALORE committed:
s7

EVALORE committed:
s8

EVALORE committed:
s9

EVALORE committed:
s10

EVALORE committed:
Add auth component, user guard, and routing logic

Implemented an authentication component with a login form, user guard for route protection, and updated routing logic to handle authenticated and unauthenticated states. Adjusted storage service and store interface for user data management. Replaced the puzzle component with a router outlet in the root component for dynamic navigation.

EVALORE committed:
Add skipAuthGuard to handle unauthenticated user logic

Introduced the `skipAuthGuard` to prevent authenticated users from accessing the authentication route. The guard checks the user's login status using `LocalStorageService` and redirects logged-in users to the root path. Included unit tests to ensure the guard's correctness and updated route configurations accordingly.

EVALORE committed:
Refactor puzzle components and services for improved clarity

Reorganized and streamlined puzzle components and services to enhance maintainability and readability. Consolidated state management logic into services, introduced better reactivity with signals and observables, and improved component interactions. Added proper methods for handling game flow, hints, storage integration, and UI updates, ensuring more modular and reusable code.

EVALORE committed:
Refactor game component and service for cleaner logic.

Removed redundant `clearSource()` method and unused imports. Organized game initialization logic into separate effects, improving readability and maintainability. Adjusted properties to ensure proper immutability where needed.

EVALORE committed:
s3

EVALORE committed:
s7

EVALORE committed:
Refactor game component and service for cleaner logic.

Removed redundant `clearSource()` method and unused imports. Organized game initialization logic into separate effects, improving readability and maintainability. Adjusted properties to ensure proper immutability where needed.

EVALORE committed:
s9

EVALORE committed:
Refactor puzzle components and services for improved clarity

Reorganized and streamlined puzzle components and services to enhance maintainability and readability. Consolidated state management logic into services, introduced better reactivity with signals and observables, and improved component interactions. Added proper methods for handling game flow, hints, storage integration, and UI updates, ensuring more modular and reusable code.

EVALORE committed:
s6

EVALORE committed:
Add skipAuthGuard to handle unauthenticated user logic

Introduced the `skipAuthGuard` to prevent authenticated users from accessing the authentication route. The guard checks the user's login status using `LocalStorageService` and redirects logged-in users to the root path. Included unit tests to ensure the guard's correctness and updated route configurations accordingly.

EVALORE committed:
Add auth component, user guard, and routing logic

Implemented an authentication component with a login form, user guard for route protection, and updated routing logic to handle authenticated and unauthenticated states. Adjusted storage service and store interface for user data management. Replaced the puzzle component with a router outlet in the root component for dynamic navigation.

EVALORE committed:
s2

EVALORE committed:
s8

EVALORE committed:
s5

EVALORE committed:
s4

EVALORE committed:
s10

EVALORE committed:
Merge branch 'new-version' of github.com:EVALORE/lingua-puzzle into new-version

EVALORE committed:
Remove unused LevelService and RoundService.

The LevelService and RoundService, along with their tests, were deleted as they were no longer used in the application. References to these services were also removed to clean up dependencies in the PuzzleComponent and related files. This improves code clarity and reduces unnecessary complexity.

intelligentRaji commented on src/app/core/guards/is-user/is-user.guard.ts:
Че-то ты напутал тут с гвардами. Сделай один гвард `auth` и все

intelligentRaji commented on src/app/auth/auth.component.ts:
```diff
@@ -0,0 +1,52 @@
+  Validators.required,
+  Validators.minLength(3),
+  Validators.pattern(/^[A-Z][a-zA-Z-]*$/u),
+];
```
```ts
const namingReallityValidators = [
  Validators.required,
  Validators.minLength(3),
  Validators.pattern(/^[A-Z][a-zA-Z-]*$/u),
];
```

intelligentRaji commented on src/app/auth/auth.component.ts:
```diff
@@ -0,0 +1,52 @@
+})
+export class AuthComponent {
+  private readonly router = inject(Router);
+  private readonly fb = inject(FormBuilder);
```
`inject(NonNullableFormBuilder)`

intelligentRaji commented on src/app/auth/auth.component.ts:
```diff
@@ -0,0 +1,52 @@
+    this.login(this.loginForm.getRawValue());
+  }
+
+  private login(data: Omit<LocalStorageStore['user'], 'progress'>): void {
```
В такой ситуации, надо чтобы `LocalStorageStore` состоял из типа `User` а не наоборот

intelligentRaji commented on src/app/auth/auth.component.ts:
```diff
@@ -0,0 +1,52 @@
+      progress: {},
+    });
+    void this.router.navigate(['']);
+    console.log(this.localStorage.getItem('user'));
```
console.log

intelligentRaji commented on src/app/puzzle/components/card-list/card-list.component.html:
```diff
@@ -0,0 +1,21 @@
+  @for (card of list().cards; track card.originalIndex) {
+    <mat-card
+      [style]="getCardStyles(card)"
+      [class]="getCardClassNames(card)"
```
Функции в темплейте вызываются на каждый CD. Тут у тебя перформанс падает

intelligentRaji commented on src/app/puzzle/components/card-list/card-list.component.ts:
Сделать отдельный класс для карточки. Перенести туда логику по созданию карты. Стили и классы сделать `computed` вместо функций

intelligentRaji commented on src/app/puzzle/components/card-list/card-list.component.html:
```diff
@@ -0,0 +1,21 @@
+      {{ card.word }}
+    </mat-card>
+  }
+}
```
Вместо того чтобы создавать два разных дерева. Ты мог бы просто сделать проверку в хендлере на состояние листа

intelligentRaji commented on src/app/puzzle/components/game/game.component.html:
```diff
@@ -0,0 +1,35 @@
+      [list]="{ type: 'result', cards: result }"
+      (cardClick)="moveCardToSource($event)"
+    />
+  }
```
`result` всегда - `true`. Убреи `@if`

intelligentRaji commented on src/app/puzzle/components/game/game.component.html:
```diff
@@ -0,0 +1,35 @@
+    [list]="{ type: 'source', cards: source }"
+    (cardClick)="moveCardToResult($event)"
+  />
+}
```
`source` всегда - `true`. Убреи `@if`

intelligentRaji commented on src/app/puzzle/services/game/game.service.ts:
```diff
@@ -0,0 +1,75 @@
+
+  public readonly sentenceIndex = signal<number>(DEFAULT_SENTENCE_INDEX);
+  public readonly completedSentences = signal<Card[][]>([]);
+  public readonly source = signal<Card[]>([]);
```
Это поле никак не влияет на состояние самой игры в целом. Поэтому возникает вопрос - а нужно ли оно в этом сервисе?

Это поле тут находиться лишь для удобства перемещения карточек. А само перемещение собственно осуществляется в компоненте игры. Так что, я думаю, это поле должно существовать только в компоненте игры

intelligentRaji commented on src/app/puzzle/services/game/game.service.ts:
```diff
@@ -0,0 +1,75 @@
+const DEFAULT_SENTENCE_INDEX = 0;
+
+@Injectable()
+export class GameService {
```
Этот сервис может зависеть от `PuzzleService`. Тогда тебе не придется запариваться с передачей карточек в этот сервис

intelligentRaji commented on src/app/puzzle/services/game/game.service.ts:
```diff
@@ -0,0 +1,75 @@
+export class GameService {
+  private readonly cardService = inject(CardService);
+
+  public readonly sentenceIndex = signal<number>(DEFAULT_SENTENCE_INDEX);
```
`setntenceIndex` поменяй на `currentSentence`

intelligentRaji commented on src/app/puzzle/services/game/game.service.ts:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  public autocompleteResult(): void {
+    this.result.update(this.cardService.sortCardsByOriginalIndex.bind(this));
```
Зачем `bind`?

intelligentRaji commented on src/app/puzzle/services/card/card.service.ts:
Зачем для этих функций сервис?

intelligentRaji commented on src/app/puzzle/services/http-data/http-data.service.ts:
```diff
@@ -0,0 +1,38 @@
+  return `wordCollectionLevel${String(levelIndex)}.json`;
+}
+
+const baseUrl = 'https://evalore.github.io/rss-puzzle-data';
```
лучше в env перенеси

intelligentRaji commented on src/app/puzzle/services/http-data/http-data.service.ts:
```diff
@@ -0,0 +1,38 @@
+
+  public getAudioFullPath(audioName: string): string {
+    return `${audioFolderUrl}/${audioName}`;
+  }
```
Этот метод не должен находиться в этом сервисе

intelligentRaji commented on src/index.html:
```diff
@@ -0,0 +1,22 @@
+      type="image/x-icon"
+      href="favicon.ico"
+    />
+      <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500&display=swap" rel="stylesheet">
```
зачем здесь таб?

intelligentRaji commented on src/app/puzzle/puzzle.component.ts:
```diff
@@ -0,0 +1,71 @@
+    })),
+    tap((data) => {
+      this.updateGameStateInLocalStorage(data.round, data.level);
+    }),
```
Эта логика должна быть в puzzle сервисе

intelligentRaji commented on src/app/puzzle/puzzle.component.ts:
```diff
@@ -0,0 +1,71 @@
+    this.updateCurrent();
+  }
+
+  private updateGameStateInLocalStorage(round: number, level: number): void {
+    this.localStorage.setItem('game', { level, round, sentenceIndex: 0 });
+  }
+
+  private updateCurrent(): void {
+    const defaultValue = {
+      round: DEFAULT_ROUND_INDEX,
+      level: DEFAULT_LEVEL,
+      sentenceIndex: 0,
+    };
+    const { level, round } = this.localStorage.getItem('game', defaultValue);
+    this.roundChange(round);
+    this.levelChange(level);
+  }
```
Перемести эту логику в сервис

intelligentRaji commented on src/app/puzzle/components/settings/settings.component.ts:
Зачем этот компонент?

intelligentRaji commented on src/app/puzzle/components/hints/hints.component.ts:
```diff
@@ -0,0 +1,43 @@
+import { MatIcon } from '@angular/material/icon';
+import { takeUntilDestroyed, toObservable } from '@angular/core/rxjs-interop';
+
+interface HintsToShow {
+  audio?: string;
+  translation?: string;
+}
```
странное название интерфейса

intelligentRaji commented on src/app/puzzle/components/hints/hints.component.ts:
```diff
@@ -0,0 +1,43 @@
+    this.showTranslation = !this.showTranslation;
+  }
+
+  public setAudioSrc(src?: string): void {
+    if (!src) {
+      return;
+    }
+    this.audio.pause();
+    this.audio = new Audio(src);
+    this.audio.load();
+  }
```
ГЕНЕАЛЬНАЯ ЛОГИКА. Сделаем src опциональным, чтобы потом проверять: "А пришло ли нам src?"

intelligentRaji commented on src/app/puzzle/components/hints/hints.component.ts:
```diff
@@ -0,0 +1,43 @@
+  protected showTranslation = false;
+
+  constructor() {
+    toObservable(this.hints)
+      .pipe(takeUntilDestroyed())
+      .subscribe((value) => {
+        this.setAudioSrc(value.audio);
+      });
```
`effect`?

intelligentRaji commented on src/app/puzzle/components/game/game.component.ts:
```diff
@@ -0,0 +1,131 @@
+
+  private numberOfSentences = computed(() => this.sentences().length);
+  protected isSourceEmpty = computed(() => !this.source().length);
+  protected isResultCorrect = computed(() =>
+    this.result().every((card) => card.positionStatus === PositionStatus.CORRECT),
+  );
+  protected isSolved = computed(
+    () => this.gameService.sentenceIndex() === this.numberOfSentences() - 1,
+  );
```
Эта логика должна быть в сервисе

intelligentRaji commented on src/app/puzzle/components/game/game.component.ts:
```diff
@@ -0,0 +1,131 @@
+  protected getBoardStyles(): BoardStyles {
+    return {
+      width: puzzleWidth.px,
+      height: `${String(cardHeight.number * this.numberOfSentences())}px`,
```
Жестка!!!

intelligentRaji commented on src/app/puzzle/components/game/game.component.ts:
```diff
@@ -0,0 +1,131 @@
+    this.initializeSourceEffect();
+  }
+
+  private initializeSentencesEffect(): void {
+    effect(() => {
+      this.sentences();
+      this.gameService.clearAll();
+    });
+  }
+
+  private initializeSourceEffect(): void {
+    effect(() => {
+      this.setSource(this.sentence().textExample);
+    });
+  }
```
Этих эффектов тут быть не должно. Это все должно быть в сервисе

intelligentRaji reviewed (approved):
Старайся!

intelligentRaji reviewed (changes requested):
Старайся!

EVALORE committed:
Integrate environment variable management with dotenv

Replaced static URLs and constants with environment variables managed via dotenv for improved configurability. Updated services, components, and type definitions to align with the new structure, and switched to the @ngx-env/builder for builds.

EVALORE committed:
Add support for environment variables handling

Introduced a consistent structure for managing environment variables across the app. Implemented `import.meta.env`, `_NGX_ENV_`, and `process.env` approaches for flexibility. Updated environment files, configuration in `angular.json`, and adjusted services for the new setup.

EVALORE committed:
Refactor data folder URLs and rename "round" to "puzzle"

Updated environment configuration to use explicit folder URLs for levels, audio, and images. Replaced "round" terminology with "puzzle" across components, services, and types for improved clarity and consistency. Removed unused interfaces and updated data flow logic in the puzzle service.

EVALORE committed:
Update Angular and related dependencies to latest versions

Upgraded Angular packages and dependencies to keep the project up-to-date. This includes improvements and bug fixes from the latest versions, ensuring compatibility across packages. Added support for TailwindCSS v4 while updating several dev dependencies.

EVALORE committed:
Refactor round logic to puzzle-based structure.

Replaces "round" terminology and logic with "puzzle" for consistency and clarity. Updates services, components, and local storage interactions accordingly. Simplifies state management and improves alignment with future scalability needs.

EVALORE committed:
Refactor round-related terminology to puzzle-specific naming

Updated variable and method names across components and services to replace "round" with "puzzle" for clarity and consistency with domain terminology. This includes renaming inputs, outputs, observables, and constants to better represent their purpose.

EVALORE committed:
Refactor puzzle state management to use a unified observable.

Replaced separate BehaviorSubjects with a single `gameState$` observable for managing puzzle state. Updated related logic and components to use the unified state, improving code readability and maintainability. Also fixed a naming inconsistency in the `default-values` import.

EVALORE committed:
Refactor CardList to Tiles for improved consistency

Renames CardList components, services, and types to Tiles for better alignment with domain terminology. Updates related logic, templates, and styles to reflect the new naming and streamline the code structure.

EVALORE committed:
Refactor guards and improve game component architecture

Replaced existing guards with `appGuard` and `authGuard` for better modularity and consolidated `isLogged` logic. Refactored the game component to enhance reusability and simplify state management by centralizing tile-related logic in `GameService`. Removed unused `CardService` and adjusted relevant components and utilities accordingly.

EVALORE committed:
Rename puzzle & game component, services, and utilities

EVALORE committed:
Rename puzzle & game component, services, and utilities

EVALORE committed:
Add modal and result components for improved user flow

Introduce reusable `ModalComponent` and `ModalService` to handle dynamic modal functionality. Implement `ResultComponent` to display puzzle results and integrate it into the game flow via the modal. Update puzzle logic and layout for better interactivity and engagement.

EVALORE committed:
Refactor audio handling and enhance UI components

Removed redundant `fullAudioPath` calls and introduced `AudioDirective` for centralized audio handling. Updated modal and result components with Material design standards and added modal actions support. Improved visual styling and accessibility for hints, statistics, and image components.

EVALORE committed:
Refactor puzzle UI and enhance game flow logic

Refactored HTML structure and SCSS for improved styling consistency and flexibility, including adjustments for puzzle actions and buttons. Introduced `isPuzzleSolved` and `showButtons` for better state management and conditional rendering. Updated global styles and added Material button module for streamlined UI components.

EVALORE committed:
Update puzzle-selector layout to use flex display

Replaced block display with flex and added a gap property. This improves layout alignment and spacing for better UI consistency.

EVALORE committed:
Refactor folder structure

Implement tsconfig paths

EVALORE committed:
Add header component with user display and logout functionality

Introduced a new `HeaderComponent` to display user information and enable logout functionality. Updated `AppComponent` to include the header and adjusted `LocalStorageService` method signatures for stricter typing. Additionally, a minor adjustment was made to `Game` type to replace `sentenceIndex` with `wordIndex`.

EVALORE committed:
Change `sentenceIndex` to `wordIndex` in GameState

Replaced all occurrences of `sentenceIndex` with `wordIndex` in the `GameState` interface and related method implementations. This update aligns the state naming with the intended functionality and improves code clarity.

EVALORE committed:
Refactor game state management and enhance header logic

Centralized game state management into GameStateService, removing redundancy in GameService. Updated header component to reflect logged-in state in the UI. Adjusted related services, components, and tests accordingly.

EVALORE committed:
Refactor and modularize user and game state management

Moved user state logic to `UserStateService` and refactored `AuthService` for centralized user authentication. Improved game components by introducing `Puzzle` and `WordEntry` models for better type integration. Consolidated UI constants and added dynamic full image handling for tiles and puzzles. Removed `HeaderComponent` in favor of inlined header logic.

EVALORE committed:
Update button label for next round in game component

Renamed the button label from "next button" to "next puzzle" for improved clarity and user experience. This ensures the label more accurately reflects its functionality.

EVALORE committed:
Add production environment configuration and update build settings

Introduced `environment.production.ts` with dynamic folder URLs for levels, audio, and images. Updated `angular.json` to include `baseHref` and `deployUrl` for production builds.
